hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclHelper.java

```diff
@@ -0,0 +1,745 @@
+        System.currentTimeMillis() - start);
+      return true;
+    } catch (Exception e) {
+      LOG.error("Set HDFS acl error when truncate {}", tableName != null ? tableName : null, e);
+      return false;
+    }
+  }
+
+  /**
+   * Remove table acl from ns dir when delete table
+   * @param tableName the table
+   * @param removeUsers the users whose acl will be removed
+   * @return false if an error occurred, otherwise true
+   */
+  public boolean removeNamespaceAcl(TableName tableName, Set<String> removeUsers) {
+    try {
+      long start = System.currentTimeMillis();
+      List<AclEntry> aclEntries =
+          removeUsers.stream().map(removeUser -> aclEntry(ACCESS, removeUser, READ_EXECUTE))
+              .collect(Collectors.toList());
+      Path nsPath = pathHelper.getDataNsDir(tableName.getNamespaceAsString());
+      fs.removeAclEntries(nsPath, aclEntries);
+      LOG.info("Remove HDFS acl when delete table {}, cost {} ms", tableName,
+        System.currentTimeMillis() - start);
+      return true;
+    } catch (Exception e) {
+      LOG.error("Set HDFS acl error when delete table {}", tableName != null ? tableName : null, e);
+      return false;
+    }
+  }
+
+  /**
+   * Set table owner acl when create table
+   * @param tableName the table
+   * @param user the table owner
+   */
+  public void addTableAcl(TableName tableName, String user) {
+    try {
+      long start = System.currentTimeMillis();
+      List<AclEntry> aclEntries = new ArrayList<>(2);
+      AclEntry accessAclEntry = aclEntry(ACCESS, user, READ_EXECUTE);
+      aclEntries.add(accessAclEntry);
+      aclEntries.add(aclEntry(DEFAULT, user, READ_EXECUTE));
+      // set access and default HDFS acl for table dir
+      fs.modifyAclEntries(pathHelper.getTmpTableDir(tableName), aclEntries);
+      fs.modifyAclEntries(pathHelper.getArchiveTableDir(tableName), aclEntries);
+      Path tableDir = pathHelper.getDataTableDir(tableName);
+      HDFSAclOperation operation = new HDFSAclOperation(fs, tableDir, Sets.newHashSet(user),
+          HDFSAclOperation.OperationType.MODIFY, READ_EXECUTE, true, new ArrayList<>(0));
+      setSingleHDFSAcl(operation).get();
+      // set access HDFS acl for ns dir
+      List<Path> nsPath = getDefaultNamespacePath(tableName.getNamespaceAsString(), false);
+      for (Path path : nsPath) {
+        fs.modifyAclEntries(path, Lists.newArrayList(accessAclEntry));
+      }
+      LOG.info("Set HDFS acl when create table {}, cost {} ms", tableName,
+        System.currentTimeMillis() - start);
+    } catch (Exception e) {
+      LOG.error("Set HDFS acl error when create table {}", tableName != null ? tableName : null, e);
+    }
+  }
+
+  /**
+   * Generate the {@link HDFSAclOperation} list
+   * @param userPermission the user and permission
+   * @param operationType MODIFY or REMOVE HDFS acl
+   * @param skipNamespaces a namespace list whose related paths will be skipped when set HDFS acl
```

Comment:
Why use an exclusive set ?  why not an inclusive set ?  It's strange here...

Reply:
There are some cases which would use a exclusive set, such as:
1. revoke user permission from ns but user owns permission of the ns table, we should remove acl of the ns directories but skip directories of this table.
2. grant user ns permission and user owns permission of the ns table, we can skip set table acls because it's already set.